espressif/vscode-esp-idf-extension
# Code coverage (VSC-312)

Add code coverage with `gcovr` using `xtensa-eslp-elf-gcov` from Xtensa Toolchain.

![Code Editor Coverage](https://user-images.githubusercontent.com/6033246/81382655-4f721600-9141-11ea-8b45-d99b5bec329f.png)

![HTML Reports](https://user-images.githubusercontent.com/6033246/81383282-6402de00-9142-11ea-8144-71b3194de0c8.png)

There is text editor line highlight for covered/uncovered lines and project gcovr html report within vscode.

Coverage files are generated on those components which has the coverage compile option enabled in CMakeLists.txt. 

This could become messy if we want to include some components source coverage as well.

Steps to test:

1) Run `ESP-IDF: Configure extension` to install gcovr into Python virtual env. `gcovr` is added in Debug Adapter's requirements.txt
2) Create a new project using coverage. For ease, just create a project using the `ESP-IDF gcov example`. [Set up ESP-IDF for App-trace Coverage](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/app_trace.html#setting-up-a-project-for-gcov)
3) Build the project and make sure there is no errors.
4) Connect to openocd server and execute `esp32 gcov dump` to generate data coverage files.
4) Run `ESP-IDF: Add Editor coverage` with source files opened in vscode editor.
5) Run `ESP-IDF: Remove editor coverage` to remove coverage highlight.
6) Run `ESP-IDF: Get HTML Coverage Report for project` to show html report in webview.
